### PR TITLE
Upgrade dependencies

### PR DESCRIPTION
**Pull request by nlebrun-spotify**
BREAKING CHANGE: TypeScript bump from v4 to v5, Jest bump from v28 to v29

Upgraded most of the dependencies, except the ones that start publishing ESM-only versions since I didn't wanted to revamp the build system to support these.

It also solve a security issue with `yaml` which had an outdated version required by some outdated dependencies.

### CODE REVIEW

File: packages/web-scripts/package.json [hunk 1]
```diff
@@ -30,34 +30,32 @@
...
+    "jest-junit": "^16.0.0",
+    "lint-staged": "^13.2.2",
+    "prettier": "^2.8.8",
+    "semantic-release": "^21.0.2",
+    "ts-jest": "^29.1.0",
+    "typescript": "^5.0.4"
   },
   "devDependencies": {
     "@types/rimraf": "^3.0.0",
-    "@types/tempy": "^0.3.0",
```
Comment (nlebrun-spotify) [inline packages/web-scripts/package.json:60]: `tempy` already provides its own types

**Comment by javoire** [comment]
Great! We'll release this as a major version of web-scripts to highlight the breaking changes

**Comment by javoire** [comment]
@nlebrun-spotify could [you set this](https://github.com/spotify/web-scripts/actions/runs/4983574825/workflow?pr=1128#L18) to `[18]` ? I don't think I have write access to your fork

**Comment by nlebrun-spotify** [comment]
> Great! We'll release this as a major version of web-scripts to highlight the breaking changes

Yep, I added the semantic-release `BREAKING CHANGE` keyword in the commit message so it will run as a major release

**Comment by nlebrun-spotify** [comment]
> We might as [well add this](https://github.com/semantic-release/semantic-release/blob/master/package.json#L78-L80) to our [own package.json as well](https://github.com/spotify/web-scripts/blob/master/packages/web-scripts/package.json#L69) to make that requirement clearer to our users

I upgrade every reference of Node version to 18, very good catch!
Also changed the CI to run tests only on Node 18 (it was throwing in lower version due to `semantic-release` requirements)

**Pull request merged**